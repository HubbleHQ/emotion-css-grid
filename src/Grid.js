/** @jsx jsx */
import React from 'react';
import PropTypes from 'prop-types';
// import { css, cx } from 'react-emotion';
import { css, jsx } from '@emotion/core';
import config from './GridConfig';

const {
  gridMaxWidth,
  columnNumber,
  gutterWidth,
  mobileBreakpoint,
} = config;

const columnWidthPX = gridMaxWidth / columnNumber;
const columnWidth = (columnWidthPX / gridMaxWidth) * 100;

const GridContainerStyles = css`
  width: 100%;
  max-width: ${gridMaxWidth}px;
  min-height: 1px;
`;
const GridContainer = ({ children, className }) => (
  <div css={GridContainerStyles} className={className}>
    {children}
  </div>
);

GridContainer.defaultProps = {
  className: '',
};

GridContainer.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]),
};

const GridStyles = css`
  width: 100%;
  max-width: ${gridMaxWidth}px;
`;
const Grid = ({ children, className }) => (
  <div css={GridStyles} className={className}>
    {children}
  </div>
);

Grid.defaultProps = {
  className: '',
};

Grid.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]),
};


const RowStyles = css`
  &:before,
  &:after {
    content: '';
    display: table;
    clear: both;
  }
`;
const RowMismatchedHeights = css`display: block;`;
const RowEqualHeights = css`
  display: block;
  @media (min-width: ${mobileBreakpoint}px) {
    display: flex;
  }
`;

const Row = ({ allowMismatchedHeights, children, className }) => (
  <div
    className={className}
    css={[
      RowStyles,
      allowMismatchedHeights ? RowMismatchedHeights : RowEqualHeights,
    ]}
  >
    {children}
  </div>
);

Row.defaultProps = {
  allowMismatchedHeights: false,
  className: '',
};

Row.propTypes = {
  allowMismatchedHeights: PropTypes.bool,
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]),
};

/* @param {number} column width expressed as a percentage
 * The responsiveWidth CSS describes how columns should
 * always collapse to a single-column layout (100% width)
 * on mobile sizes. At any size above a mobile range,
 * the column's width will be set by whatever numeric
 * percentage value argument that is passed in.
 */
const responsiveWidth = widthValue => css`
  width: 100%;
  @media (min-width: ${mobileBreakpoint}px) {
    width: ${widthValue}%;
  }
`;

/*
 * We omit gutters on columns by default on mobile sizes
 */
const responsiveGutters = css`
  padding: 0;
  @media (min-width: ${mobileBreakpoint}px) {
    padding-left: ${gutterWidth / 2}px;
    padding-right: ${gutterWidth / 2}px;
  }
`;

const ColumnStyles = css`
  ${responsiveGutters};
  ${responsiveWidth(columnWidth)};
  float: left;
  min-height: 1px;
`;
const Column = ({ children, className }) => (
  <div css={ColumnStyles} className={className}>
    {children}
  </div>
);

Column.defaultProps = {
  className: '',
};

Column.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]),
};

/*
 * @param {number} column width expressed as a whole number from 1 to max Columns
 * @returns {(string|null)} EmotionJS CSS string, or null
 * @example
 * <Column css={Col(2)}>
 * <Column css={[Col(6), someOtherStyling]}>
 * NOTE: just because you CAN add additional styling to a column doesn't mean you SHOULD.
 */
const Col = (number) => {
  if (
    number > columnNumber
    || number < 1
    || typeof number !== 'number'
  ) {
    // eslint-disable-next-line no-console
    console.error(
      `This is a ${
        columnNumber
      } column grid. The value supplied to Col (${number}) won't work!`,
    );
    return null;
  }
  return css`
    ${responsiveWidth(columnWidth * number)};
  `;
};

export {
  GridContainer, Grid, Row, Column, Col,
};
