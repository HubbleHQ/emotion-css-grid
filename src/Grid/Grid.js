/** @jsx jsx */
import PropTypes from 'prop-types';
import { css, jsx } from '@emotion/core';
import config from './GridConfig';

const { gridMaxWidth } = config;

const gridStyles = css`
  width: 100%;
  max-width: ${gridMaxWidth}px;
`;
const Grid = ({ children, className }) => (
  <div css={gridStyles} className={className}>
    {children}
  </div>
);

Grid.defaultProps = {
  className: '',
};

Grid.propTypes = {
  /**
   * Grid elements: should be <Row />.
   */
  children: PropTypes.node.isRequired,
  /**
   * Optional class name.
   */
  className: PropTypes.string,
};

export default Grid;