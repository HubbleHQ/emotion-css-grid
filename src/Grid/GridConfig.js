
/**
 * columnNumber: total number of columns in the grid, expressed as an integer
 * gridMaxWidth: total width of grid expressed in pixels
 * gutterWidth: horizontal padding on columns, expressed in pixels
 * mobileBreakpoint: the screen width (in pixels) at which the grid
 * stops being laid out for mobile and starts being laid out for desktop
 */
export default {
  columnNumber: 12,
  gridMaxWidth: 1920,
  gutterWidth: 20,
  mobileBreakpoint: 640,
};
