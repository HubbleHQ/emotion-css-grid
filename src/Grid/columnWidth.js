import config from './GridConfig';

const { columnNumber } = config;

/*
 * Column width is expressed as a percentage value
 * based on the total number of columns specified in grid config.
 */
const columnWidth = 100 / columnNumber;

export default columnWidth;
