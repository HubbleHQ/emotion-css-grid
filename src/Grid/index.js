export { default as Grid } from './Grid';
export { default as Row } from './Row';
export { default as Column } from './Column';
export { default as col } from './col';
export { default as columnWidth } from './columnWidth';
export { default as responsiveWidth } from './responsiveWidth';
