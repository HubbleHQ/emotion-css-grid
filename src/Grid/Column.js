/** @jsx jsx */
import PropTypes from 'prop-types';
import { css, jsx } from '@emotion/core';

import config from './GridConfig';
import columnWidth from './columnWidth';
import responsiveWidth from './responsiveWidth';

const { gutterWidth, mobileBreakpoint } = config;

/*
 * We omit gutters on columns by default on mobile sizes
 */
const responsiveGutters = css`
  padding: 0;
  @media (min-width: ${mobileBreakpoint}px) {
    padding-left: ${gutterWidth / 2}px;
    padding-right: ${gutterWidth / 2}px;
  }
`;

const columnStyles = css`
  ${responsiveGutters};
  ${responsiveWidth(columnWidth)};
  float: left;
  min-height: 1px;
`;

const Column = ({ children, className }) => (
  <div css={columnStyles} className={className}>
    {children}
  </div>
);

Column.defaultProps = {
  className: '',
};

Column.propTypes = {
  /**
   * Column elements: this is where the page content will sit,
   * so these elements can be headings, <div> etc.
   */
  children: PropTypes.node.isRequired,
  /**
   * Optional class name.
   */
  className: PropTypes.string,
};
