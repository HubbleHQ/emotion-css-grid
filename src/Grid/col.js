import { css } from '@emotion/core';
import config from './GridConfig';
import columnWidth from './columnWidth';
import responsiveWidth from './responsiveWidth';

const { columnNumber } = config;

/*
 * @param {number} column width expressed as a whole number from 1 to max Columns
 * @returns {(string|null)} EmotionJS CSS string, or null
 * @example
 * <Column css={col(2)}>
 * <Column css={[col(6), someOtherStyling]}>
 * NOTE: just because you CAN add additional styling to a column doesn't mean you SHOULD.
 */
const col = (number) => {
  if (
    number > columnNumber
    || number < 1
    || typeof number !== 'number'
  ) {
    throw new Error(`This is a ${columnNumber} column grid. 
    The value supplied to col (${number}) won't work!`);
  }
  return css`
    ${responsiveWidth(columnWidth * number)};
    `;
};

export default col;
