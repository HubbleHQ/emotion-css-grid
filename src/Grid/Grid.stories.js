/** @jsx jsx */
import { storiesOf } from '@storybook/react';

import { css, jsx } from '@emotion/core';

import Grid from './Grid';
import Row from './Row';
import Column from './Column';
import col from './col';

const gridStyle = css`
  border: 1px solid black;
  background-color: #eee;
`;

const rowStyle = css`
  border-top: 1px solid black;

  &:first-child {
    border-top: none;
  }

  &:nth-child(odd) {
    background-color: #ddd;
  }
`;

const body = css`
  font-family: sans-serif;
  font-size: 16px;
  font-weight: lighter;
`;

storiesOf('Emotion CSS Grid', module)
  .add('Default', () => (
    <div css={body}>
      <h3>Our Grid</h3>
      <p>
        The markup in these storybook views assume default values in the config file.
      </p>
      <p>
        <strong>
          If you change the values in the config file, please expect the storybook demo
          visuals to change as well.
        </strong>
      </p>
      <p>
        At mobile sizes, we expect these columns to become 100% width so we have
        a single-column layout. If you&apos;re at a tiny screen size and everything
        looks like it&apos;s all 100% wide, that&apos;s the correct behaviour :)
      </p>
      <Grid css={gridStyle}>
        <Row css={rowStyle}>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
          <Column>1/12th</Column>
        </Row>
        <Row css={rowStyle}>
          <Column css={col(2)}>1/6th</Column>
          <Column css={col(2)}>1/6th</Column>
          <Column css={col(2)}>1/6th</Column>
          <Column css={col(2)}>1/6th</Column>
          <Column css={col(2)}>1/6th</Column>
          <Column css={col(2)}>1/6th</Column>
        </Row>
        <Row css={rowStyle}>
          <Column css={col(4)}>
            <p>This column is the width of 4 columns (33%)!</p>
          </Column>
          <Column css={col(4)}>
            <p>This column is the width of 4 columns (33%)!</p>
          </Column>
          <Column css={col(4)}>
            <p>This column is the width of 4 columns (33%)!</p>
          </Column>
        </Row>
        <Row css={rowStyle}>
          <Column css={col(6)}>
            <p>This column is the width of 6 columns (50%)!</p>
          </Column>
          <Column css={col(6)}>
            <p>This column is the width of 6 columns (50%)!</p>
          </Column>
        </Row>
        <Row css={rowStyle}>
          <Column css={col(3)}>
            <p>This column is the width of 3 columns!</p>
          </Column>
          <Column css={col(3)}>
            <p>This column is the width of 3 columns!</p>
          </Column>
          <Column css={col(3)}>
            <p>This column is the width of 3 columns!</p>
          </Column>
          <Column css={col(3)}>
            <p>This column is the width of 3 columns!</p>
          </Column>
        </Row>
        <Row css={rowStyle}>
          <Column css={col(3)}>
            <p>This column is the width of 3 columns!</p>
          </Column>
          <Column css={col(3)}>
            <p>This column is the width of 3 columns!</p>
          </Column>
          <Column css={col(6)}>
            <p>This column is the width of 6 columns!</p>
          </Column>
        </Row>
        <Row css={rowStyle}>
          <Column css={col(12)}>
            <p>This column is the width of 12 columns!</p>
          </Column>
        </Row>
      </Grid>
    </div>
  ))
  .add('Equal Heights Toggle', () => (
    <div css={body}>
      <Grid css={gridStyle}>
        <Row css={rowStyle}>
          <Column css={col(6)}>
            <p>
            This row is using the&nbsp;
              <strong>default style rules</strong>
              &nbsp;which apply flex in order to have equal-height columns regardless of content.
            </p>
          </Column>
          <Column css={col(6)}>
            <p>This column is the width of 6 columns!</p>
            <p>
              Sherlock Holmes closed his eyes and placed his elbows upon the arms of his chair,
              with his finger-tips together. "The ideal reasoner," he remarked, "would, when he
              had once been shown a single fact in all its bearings, deduce from it not only
              all the chain of events which led up to it but also all the results which would
              follow from it.
            </p>
          </Column>
        </Row>
        <Row allowMismatchedHeights css={rowStyle}>
          <Column css={css`background: skyblue;`}>
            <p>This column is the width of 1 column (1/12th)!</p>
          </Column>
          <Column css={[col(6), css`background: pink;`]}>
            <p>
              This row has the prop &nbsp;
              <strong>allowMismatchedHeights</strong>
              &nbsp;applied, which means it will not use flexbox to equalise the
              heights of the child columns.
            </p>
            <p>
              Sherlock Holmes closed his eyes and placed his elbows upon the arms of his chair,
              with his finger-tips together. "The ideal reasoner," he remarked, "would, when he
              had once been shown a single fact in all its bearings, deduce from it not only
              all the chain of events which led up to it but also all the results which would
              follow from it.
            </p>
          </Column>
          <Column css={[col(5), css`background: yellow;`]}>
            <p>This column is the width of 5 columns!</p>
          </Column>
        </Row>
        <Row css={rowStyle}>
          <Column>
            <p>This column is the width of 1 column (1/12th)!</p>
          </Column>
          <Column css={col(6)}>
            <p>
              This row is using the default style rules which apply flex in order
              to have equal-height columns regardless of content.
            </p>
          </Column>
          <Column css={col(5)}>
            <p>This column is the width of 5 columns!</p>
          </Column>
        </Row>
      </Grid>
    </div>
  ))
  .add('Example with Content', () => {

    const webpageRowStyle = css`
      background-color: powderblue;
      padding: 15px 0;
    `;

    const backgroundImg = css`
      padding: 15px;
      background-image: url("https://picsum.photos/id/10/600/600");
      background-size: cover;
      height: 100%;

      & h2, h3, p {
        margin: 0 0 15px;
        color: white;
      }
    `;

    const imageRow = css`
      background-size: cover;
      overflow: hidden;
      height: 100px;
      min-width: 100px;
    `;

    const footerText = css`
      padding-right: 15px;
      text-align: right;
      font-weight: 400;
      color: white;
    `;
    return (
      <div css={body}>
        <Grid>
          <Row css={webpageRowStyle}>
            <Column css={col(12)}>
              <h1>Sherlock Holmes closed his eyes</h1>
            </Column>
          </Row>
          <Row>
            <Column css={col(6)}>
              <h2>And placed his elbows upon the arms of his chair</h2>
              <p>
                Problems may be solved in the study which have baffled all those who have sought
                a solution by the aid of their senses. To carry the art, however, to its highest
                pitch, it is necessary that the reasoner should be able to utilise all the facts
                which have come to his knowledge; and this in itself implies, as you will readily
                see, a possession of all knowledge, which, even in these days of free education
                and encyclopaedias, is a somewhat rare accomplishment.
              </p>
            </Column>
            <Column css={col(6)}>
              <div css={backgroundImg}>
                <h2>
                  That such a change should have taken place was incredible to him.
                </h2>
                <h3>
                  And yet it was a fact.
                </h3>
                <p>
                  As he often remembered afterwards, and always with no small wonder,
                  he found himself at first gazing at the portrait with a feeling of
                  almost scientific interest.
                </p>
              </div>
            </Column>
          </Row>
          <Row css={webpageRowStyle}>
            <Column css={col(3)}>
              <div css={[imageRow, css`background-image: url('https://picsum.photos/id/1019/400/400')`]} />
            </Column>
            <Column css={col(3)}>
              <div css={[imageRow, css`background-image: url('https://picsum.photos/id/1015/400/400')`]} />
            </Column>
            <Column css={col(3)}>
              <div css={[imageRow, css`background-image: url('https://picsum.photos/id/10/400/400')`]} />
            </Column>
            <Column css={col(3)}>
              <div css={[imageRow, css`background-image: url('https://picsum.photos/id/1002/400/400')`]} />
            </Column>
          </Row>
          <Row>
            <Column css={col(4)}>
              <h3>And yet it was a fact.</h3>
              <p>
                Problems may be solved in the study which have baffled all those who have sought
                a solution by the aid of their senses.
              </p>
            </Column>
            <Column css={col(4)}>
              <h3>And yet it was a fact.</h3>
              <p>
                To carry the art, however, to its highest pitch, it is necessary that the reasoner
                should be able to utilise all the facts which have come to his knowledge.
              </p>
            </Column>
            <Column css={col(4)}>
              <h3>And yet it was a fact.</h3>
              <p>
                This in itself implies, as you will readily see, a possession of all knowledge,
                which, even in these days of free education and encyclopaedias, is a somewhat
                rare accomplishment.
              </p>
            </Column>
          </Row>
          <Row>
            <Column css={col(12)}>
              <div css={[imageRow, css`background-image: url('https://picsum.photos/id/1002/1280/200')`]}>
                <h4 css={footerText}>A feeling of almost scientific interest.</h4>
              </div>
            </Column>
          </Row>
        </Grid>
      </div>
    );
  });
