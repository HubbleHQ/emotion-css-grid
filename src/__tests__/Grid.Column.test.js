/** @jsx jsx */
import React from 'react';
import { shallow } from 'enzyme';

import { css, jsx } from '@emotion/core';
import { Column, Col } from '../Grid';

const mockContent = <p>Hello World</p>;

describe('Grid : Column', () => {
  it('should render as expected in its default state', () => {
    const output = shallow(<Column>{mockContent}</Column>);
    expect(output).toMatchSnapshot();
  });
  it('should render as expected when passed a Col(n) method as a className', () => {
    const output = shallow(<Column css={Col(4)}>{mockContent}</Column>);
    expect(output).toMatchSnapshot();
  });
  it('should render as expected when passed a Col(n) method and other styling', () => {
    const output = shallow(
      <Column
        css={[
          Col(2),
          css`
            border: 1px solid blue;
          `,
        ]}
      >
        {mockContent}
      </Column>,
    );
    expect(output).toMatchSnapshot();
  });
  it('should allow additional styling to override Col(n) method rules', () => {
    const output = shallow(
      <Column
        css={[
          Col(6),
          css`
            width: 50%;
            @media (min-width: 768px) {
              width: 75%;
            }
          `,
        ]}
      >
        {mockContent}
      </Column>,
    );
    expect(output).toMatchSnapshot();
  });
});
