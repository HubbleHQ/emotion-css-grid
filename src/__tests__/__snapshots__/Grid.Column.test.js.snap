// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid : Column should allow additional styling to override Col(n) method rules 1`] = `
.emotion-0 {
  padding: 0;
  width: 100%;
  float: left;
  min-height: 1px;
  width: 100%;
  width: 50%;
}

@media (min-width:640px) {
  .emotion-0 {
    padding-left: 10px;
    padding-right: 10px;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 8.333333333333332%;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 49.99999999999999%;
  }
}

@media (min-width:768px) {
  .emotion-0 {
    width: 75%;
  }
}

<div
  className="emotion-0"
>
  <p>
    Hello World
  </p>
</div>
`;

exports[`Grid : Column should render as expected in its default state 1`] = `
.emotion-0 {
  padding: 0;
  width: 100%;
  float: left;
  min-height: 1px;
}

@media (min-width:640px) {
  .emotion-0 {
    padding-left: 10px;
    padding-right: 10px;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 8.333333333333332%;
  }
}

<div
  className="emotion-0"
>
  <p>
    Hello World
  </p>
</div>
`;

exports[`Grid : Column should render as expected when passed a Col(n) method and other styling 1`] = `
.emotion-0 {
  padding: 0;
  width: 100%;
  float: left;
  min-height: 1px;
  width: 100%;
  border: 1px solid blue;
}

@media (min-width:640px) {
  .emotion-0 {
    padding-left: 10px;
    padding-right: 10px;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 8.333333333333332%;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 16.666666666666664%;
  }
}

<div
  className="emotion-0"
>
  <p>
    Hello World
  </p>
</div>
`;

exports[`Grid : Column should render as expected when passed a Col(n) method as a className 1`] = `
.emotion-0 {
  padding: 0;
  width: 100%;
  float: left;
  min-height: 1px;
  width: 100%;
}

@media (min-width:640px) {
  .emotion-0 {
    padding-left: 10px;
    padding-right: 10px;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 8.333333333333332%;
  }
}

@media (min-width:640px) {
  .emotion-0 {
    width: 33.33333333333333%;
  }
}

<div
  className="emotion-0"
>
  <p>
    Hello World
  </p>
</div>
`;
